server/migrator/Exporter.java

```diff
@@ -0,0 +1,197 @@
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.stream.Stream;
+
+import static grakn.core.common.exception.ErrorMessage.Internal.ILLEGAL_STATE;
+import static grakn.core.common.exception.ErrorMessage.Migrator.FILE_NOT_WRITABLE;
+
+public class Exporter implements Migrator {
+
+    private static final Logger LOG = LoggerFactory.getLogger(Exporter.class);
+    private final Grakn.Session session;
+    private final Path filename;
+    private long totalThingCount = 0;
+    private final AtomicLong entityCount = new AtomicLong(0);
+    private final AtomicLong relationCount = new AtomicLong(0);
+    private final AtomicLong attributeCount = new AtomicLong(0);
+    private final AtomicLong ownershipCount = new AtomicLong(0);
+    private final AtomicLong playerCount = new AtomicLong(0);
+
+    public Exporter(Grakn grakn, String database, Path filename) {
+        this.session = grakn.session(database, Arguments.Session.Type.DATA);
+        this.filename = filename;
+    }
+
+    @Override
+    public MigratorProto.Job.Progress getProgress() {
+        long current = attributeCount.get() + relationCount.get() + entityCount.get();
+        return MigratorProto.Job.Progress.newBuilder()
+                .setCurrent(current)
+                .setTotal(totalThingCount)
+                .build();
+    }
+
+    @Override
+    public void run() {
+        LOG.info("Exporting {} from Grakn {}", session.database().name(), Version.VERSION);
+        try (OutputStream outputStream = new BufferedOutputStream(Files.newOutputStream(filename))) {
+            try (Grakn.Transaction tx = session.transaction(Arguments.Transaction.Type.READ)) {
+                totalThingCount = tx.concepts().getRootThingType().getInstancesCount();
+                DataProto.Item header = DataProto.Item.newBuilder()
+                        .setHeader(DataProto.Item.Header.newBuilder()
+                                .setGraknVersion(Version.VERSION)
+                                .setOriginalDatabase(session.database().name()))
+                        .build();
+                write(outputStream, header);
+
+                List<Runnable> workers = new ArrayList<>();
+                workers.add(() -> tx.concepts().getRootEntityType().getInstances().forEach(entity -> {
+                    DataProto.Item item = readEntity(entity);
+                    write(outputStream, item);
+                }));
+                workers.add(() -> tx.concepts().getRootRelationType().getInstances().forEach(relation -> {
+                    DataProto.Item item = readRelation(relation);
+                    write(outputStream, item);
+                }));
+                workers.add(() -> tx.concepts().getRootAttributeType().getInstances().forEach(attribute -> {
+                    DataProto.Item item = readAttribute(attribute);
+                    write(outputStream, item);
+                }));
+                workers.parallelStream().forEach(Runnable::run);
```

Comment:
can you confirm this actually runs in parallel? I think we've seen that this sometimes is parallel and sometimes is not. May be better to use an Async executor service to guarantee this happens as desired.

Reply:
This is copied from old code, will double check.

Reply:
`parallelStream` does return a stream with `isParallel` set to true.